### PR TITLE
make and, or, xor, add - variadic

### PR DESCRIPTION
**Pull request by drom**
:construction: do not merge :construction: 

Work towards #10

### CODE REVIEW

File: include/circt/Dialect/RTL/Combinatorial.td [hunk 1]
```diff
@@ -63,6 +63,27 @@ class UTBinRTLOp<string mnemonic, list<OpTrait> traits = []> :
   }];
 }
 
+// Base class for variadic operators.
+class VariRTLOp<string mnemonic, list<OpTrait> traits = []> :
```
Comment (River707) [inline include/circt/Dialect/RTL/Combinatorial.td:67]: nit: The formatting on this seems a bit weird. Does Circt have a specific style guide for tablegen? Genuinely curious.

Comment (drom) [reply include/circt/Dialect/RTL/Combinatorial.td:67]: I don't know about CIRCT style guide for tablegen. How would you change it?

Comment (lattner) [reply include/circt/Dialect/RTL/Combinatorial.td:67]: We don't have a bespoke style guide here, we are trying to follow MLIR style for "all the things"

Comment (lattner) [reply include/circt/Dialect/RTL/Combinatorial.td:67]: I think that more typical style would be to put the colon on the next line before `RTLOp<`

**Comment by lattner** [comment]
This WIP is looking great so far @drom !

File: include/circt/Dialect/RTL/Combinatorial.td (hunk 1 above)
Comment (stephenneuendorffer) [reply include/circt/Dialect/RTL/Combinatorial.td:67]: The lack of decent autoformatting for tablegen is something I've always been annoyed by, but never quite bothered trying to fix.

File: include/circt/Dialect/RTL/Combinatorial.td [hunk 2]
```diff
@@ -63,6 +63,27 @@ class UTBinRTLOp<string mnemonic, list<OpTrait> traits = []> :
...
+  let arguments = (ins Variadic<AnySignlessInteger>:$inputs);
+  let results = (outs AnySignlessInteger:$result);
+
+  let assemblyFormat = [{
+    $inputs  attr-dict `:` functional-type($inputs, results)
+    }];
+  }
+
+// Variadic operator with uniform input/result types.
+class UTVariRTLOp<string mnemonic, list<OpTrait> traits = []> :
```
Comment (stephenneuendorffer) [inline include/circt/Dialect/RTL/Combinatorial.td:78]: Prefer full words, rather than "Vari"

File: include/circt/Dialect/RTL/Combinatorial.td (hunk 1 above)
Comment (lattner) [reply include/circt/Dialect/RTL/Combinatorial.td]: Yeah, agreed, it would be really nice.

**Comment by amaleewilson** [comment]
@drom  and I have tried a few things for debugging issues with variadic operations. Something that has been difficult to debug is a compiler error for a generated file:

```
include/circt/Dialect/RTL/RTL.cpp.inc:2673:31: error: ‘class mlir::ValueRange’ has no member named ‘front’
   odsState.addTypes({operands.front().getType()});
                               ^~~~~
include/circt/Dialect/RTL/RTL.cpp.inc:2673:49: error: no matching function for call to ‘mlir::OperationState::addTypes(<brace-enclosed initializer list>)’
   odsState.addTypes({operands.front().getType()});
```

The error-causing code:
```
void XorOp::build(OpBuilder &odsBuilder, OperationState &odsState, ValueRange operands) {
  odsState.addOperands(operands);
  odsState.addTypes({operands.front().getType()});
}
```

I changed VariadicRTLOp to look like the following in in Combinatorial.td, but @drom was seeing the same error with previous iterations:

```
class VariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
      RTLOp<mnemonic, !listconcat(traits, [NoSideEffect])> {
  let arguments = (ins Variadic<AnySignlessInteger>:$operands);
  let results = (outs AnySignlessInteger:$result);

  let assemblyFormat = [{
    `(` $operands `)`  attr-dict `:` functional-type($operands, results)
  }];
}

class UTVariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
      VariadicRTLOp<mnemonic,
               !listconcat(traits,
                           [SameTypeOperands, SameOperandsAndResultType])> {
   let assemblyFormat = [{
    `(` $operands `)`  attr-dict `:` type($result)
  }];
}
```

I also tried several other variations. When XorOp is a VariadicRTLOp (`def XorOp : UTVariadicRTLOp<"xor">;`), no problem. But when it is a UTVariadicRTLOp, I get the error. The non-problematic code (when XorOp is VariadicRTLOp and not UTVariadicRTLOp) looks like this:
```
void XorOp::build(OpBuilder &odsBuilder, OperationState &odsState, Type result, ValueRange operands) {
  odsState.addOperands(operands);
  odsState.addTypes(result);
}
```

Any ideas for fixing this? Looked through the ODS docs and other td files, but still a bit stuck. Commenting out the tabelgen'd code that causes the error doesn't seem like a good solution.

**Comment by stephenneuendorffer** [comment]
> Is the ODS wrong? Are variadic things finicky? What's up with generating the operands.front() call if ValueRange doesn't have a front()? Looked through the ODS docs and other td files, but still a bit stuck. Commenting out the tabelgen'd code that causes the error doesn't seem like a good solution.

I suspect ODS is wrong when dealing with Variadic operations and SameOperandsAndResultType.  ValueRange used to be a vector, but now its not, so I suspect this case just didn't get fixed.  It should probably be something like *(range.getTypes().begin())

**Comment by River707** [comment]
> > Is the ODS wrong? Are variadic things finicky? What's up with generating the operands.front() call if ValueRange doesn't have a front()? Looked through the ODS docs and other td files, but still a bit stuck. Commenting out the tabelgen'd code that causes the error doesn't seem like a good solution.
> 
> I suspect ODS is wrong when dealing with Variadic operations and SameOperandsAndResultType. ValueRange used to be a vector, but now its not, so I suspect this case just didn't get fixed. It should probably be something like *(range.getTypes().begin())

Yes, I don't think this wasactually tested anywhere which is unfortunate. Should be fixed in https://github.com/llvm/llvm-project/commit/6b9a706200cbb27e5e863cfd41fb3684ee616e23

**Comment by stephenneuendorffer** [comment]
> > > Is the ODS wrong? Are variadic things finicky? What's up with generating the operands.front() call if ValueRange doesn't have a front()? Looked through the ODS docs and other td files, but still a bit stuck. Commenting out the tabelgen'd code that causes the error doesn't seem like a good solution.
> > 
> > 
> > I suspect ODS is wrong when dealing with Variadic operations and SameOperandsAndResultType. ValueRange used to be a vector, but now its not, so I suspect this case just didn't get fixed. It should probably be something like *(range.getTypes().begin())
> 
> Yes, I don't think this was actually tested anywhere which is unfortunate. Should be fixed in [llvm/llvm-project@6b9a706](https://github.com/llvm/llvm-project/commit/6b9a706200cbb27e5e863cfd41fb3684ee616e23)

Thanks River!

**Comment by amaleewilson** [comment]
@stephenneuendorffer Thank you for bumping llvm. Could you bump it again to include the fix that @River707 submitted?

File: lib/Dialect/FIRRTL/LowerToRTL.cpp [hunk 3]
```diff
@@ -163,6 +163,57 @@ static LogicalResult lower(firrtl::CatPrimOp op, ArrayRef<Value> operands,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// Variadic Bitwise Operations
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
```
Comment (lattner) [inline lib/Dialect/FIRRTL/LowerToRTL.cpp:170]: Random driveby comment (please lemme know when you're ready for review, I realize this is a work in progress):

Please don't use <iostream> in CIRCT, it has several problems:
https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

File: include/circt/Dialect/RTL/Combinatorial.td [hunk 4]
```diff
@@ -63,15 +63,39 @@ class UTBinRTLOp<string mnemonic, list<OpTrait> traits = []> :
...
 def AddOp : UTBinRTLOp<"add", [Commutative]>;
 def SubOp : UTBinRTLOp<"sub">;
 def MulOp : UTBinRTLOp<"mul", [Commutative]>;
 def DivOp : UTBinRTLOp<"div">;
 def ModOp : UTBinRTLOp<"mod">;
-def AndOp : UTBinRTLOp<"and", [Commutative]>;
-def OrOp  : UTBinRTLOp<"or",  [Commutative]>;
-def XorOp : UTBinRTLOp<"xor", [Commutative]>;
+
+def AndOp : UTVariadicRTLOp<"and">;
```
Comment (lattner) [inline include/circt/Dialect/RTL/Combinatorial.td:93]: Curiosity: does MLIR's "Commutative" flag work with variadic operators?  It seems like it could/should allow canonicalization of constants to the RHS etc.  Perhaps "commutative+reassociative" is a closer match to what we need though.  That would allow turning "(and 1, x, 2)" into "(and x, (and 1, 2))" which then constant folds.

@River707 have you thought about this at all?

Comment (stephenneuendorffer) [reply include/circt/Dialect/RTL/Combinatorial.td:93]: @lattner I added support for variadic commutative operators a while back.

File: lib/Dialect/FIRRTL/LowerToRTL.cpp (hunk 3 above)
Comment (amaleewilson) [reply lib/Dialect/FIRRTL/LowerToRTL.cpp:170]: Yes, iostream was never going to be pushed; it's an artifact from the debugging pair programming, but thanks for the reminder to remove it.

**Comment by amaleewilson** [comment]
Copy-pasted River's fix to unblock variadic stuff and no longer getting the compiler error from before. However, I am still seeing some unexpected behavior. This is the variadic ops:

```
class VariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
      RTLOp<mnemonic, !listconcat(traits, [NoSideEffect])> {
  let arguments = (ins Variadic<AnySignlessInteger>:$operands);
  let results = (outs AnySignlessInteger:$result);

  let assemblyFormat = [{
    $operands  attr-dict `:` functional-type($operands, results)
  }];
}

class UTVariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
      VariadicRTLOp<mnemonic,
               !listconcat(traits,
                           [SameTypeOperands, SameOperandsAndResultType])> {
   let assemblyFormat = [{
    $operands  attr-dict `:` type($result)
  }];

def AndOp : UTVariadicRTLOp<"and", [Commutative]>;
def OrOp  : UTVariadicRTLOp<"or", [Commutative]>;
def XorOp : UTVariadicRTLOp<"xor", [Commutative]>;
```

And getting this error from a test: 
```
./bin/circt-opt ../test/rtl/bitwise.mlir -debug
Args: ./bin/circt-opt ../test/rtl/bitwise.mlir -debug 
../test/rtl/bitwise.mlir:8:22: error: expected ':'
  %and1 = rtl.and %a %b : i7
                     ^
```

From what I can tell (from this and some other experiments) it is expecting one operand, even though the and op is specified as being variadic.

**Comment by River707** [comment]
> Copy-pasted River's fix to unblock variadic stuff and no longer getting the compiler error from before. However, I am still seeing some unexpected behavior. This is the variadic ops:
> 
> ```
> class VariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
>       RTLOp<mnemonic, !listconcat(traits, [NoSideEffect])> {
>   let arguments = (ins Variadic<AnySignlessInteger>:$operands);
>   let results = (outs AnySignlessInteger:$result);
> 
>   let assemblyFormat = [{
>     $operands  attr-dict `:` functional-type($operands, results)
>   }];
> }
> 
> class UTVariadicRTLOp<string mnemonic, list<OpTrait> traits = []> :
>       VariadicRTLOp<mnemonic,
>                !listconcat(traits,
>                            [SameTypeOperands, SameOperandsAndResultType])> {
>    let assemblyFormat = [{
>     $operands  attr-dict `:` type($result)
>   }];
> 
> def AndOp : UTVariadicRTLOp<"and", [Commutative]>;
> def OrOp  : UTVariadicRTLOp<"or", [Commutative]>;
> def XorOp : UTVariadicRTLOp<"xor", [Commutative]>;
> ```
> 
> And getting this error from a test:
> 
> ```
> ./bin/circt-opt ../test/rtl/bitwise.mlir -debug
> Args: ./bin/circt-opt ../test/rtl/bitwise.mlir -debug 
> ../test/rtl/bitwise.mlir:8:22: error: expected ':'
>   %and1 = rtl.and %a %b : i7
>                      ^
> ```
> 
> From what I can tell (from this and some other experiments) it is expecting one operand, even though the and op is specified as being variadic. Again, I'm not sure if this is a bug or a me-problem.

It's parsed/printed as a list, i.e. you are missing a comma between the SSA values.

**Comment by stephenneuendorffer** [comment]
> Copy-pasted River's fix to unblock variadic stuff and no longer getting the compiler error from before.

The change is queued up... just waiting for the CI to finish.

**Comment by amaleewilson** [comment]
> It's parsed/printed as a list, i.e. you are missing a comma between the SSA values.

```
./bin/circt-opt ../test/rtl/bitwise.mlir -debug
../test/rtl/bitwise.mlir:8:19: error: custom op 'rtl.and' 2 operands present, but expected 1
  %and1 = rtl.and %a, %b : i7
```

**Comment by stephenneuendorffer** [comment]
> > Copy-pasted River's fix to unblock variadic stuff and no longer getting the compiler error from before.
> 
> The change is queued up... just waiting for the CI to finish.

OK, it's in.

**Comment by River707** [comment]
> > It's parsed/printed as a list, i.e. you are missing a comma between the SSA values.
> 
> ```
> ./bin/circt-opt ../test/rtl/bitwise.mlir -debug
> ../test/rtl/bitwise.mlir:8:19: error: custom op 'rtl.and' 2 operands present, but expected 1
>   %and1 = rtl.and %a, %b : i7
> ```

Took a look and seems like OpFormatGen doesn't support inferring a variadic from a non-variadic, should be fixed with https://reviews.llvm.org/D83011 if someone wants to take a look. I'll be OOO until next thursday starting tomorrow, so it'd need to be reviewed soon to make sure it is in before I leave.

**Comment by amaleewilson** [comment]
With the update from @River707, everything works and is otherwise ready for review. Will leave this PR as a draft, though, because these changes will break tests until the llvm version is once again bumped.

**Comment by drom** [comment]
It works E2E with https://reviews.llvm.org/D83011

**Comment by lattner** [comment]
great work everyone!

File: lib/EmitVerilog/EmitVerilog.cpp [hunk 5]
```diff
@@ -899,6 +902,17 @@ SubExprInfo ExprEmitter::emitBinary(Operation *op, VerilogPrecedence prec,
   return {prec, signedness};
 }
 
+SubExprInfo ExprEmitter::emitVariadic(Operation *op, VerilogPrecedence prec,
+                                      const char *syntax, bool hasStrictSign) {
+  int i = 0;
+  for (auto v1 : op->getOperands()) {
```
Comment (lattner) [inline lib/EmitVerilog/EmitVerilog.cpp:908]: I'd recommend using the llvm::interleave template here from llvm/ADT/STLExtras.h

**Comment by drom** [comment]
I compile fine on my machine. @tobiasgrosser do you know why CI is failing?
It seems to skip the compilation of LLVM. Is it using the old version? is it the "cache invalidation" problem?

**Comment by tobiasgrosser** [comment]
Not sure, I am on a bike tour today, but can debug this tonight.

**Comment by drom** [comment]
@tobiasgrosser looks like it taking old LLVM from the cache.

**Comment by tobiasgrosser** [comment]
> @tobiasgrosser looks like it taking old LLVM from the cache.

It seems there is an issue with submodule updates in combination with pull requests. If you check the tests, the actual CI run for your push worked flawless and is green, but the pull request somehow does not get the correct new submodule, I guess. It is late today, but I will investigate it in more depth soon. Until then, it should be safe to just check the result of the "push" run. In this case, the CI passes and consequently, it seems safe to merge from a pure CI standpoint.

**Comment by tobiasgrosser** [comment]
@drom, I just re-checked in my own fork and it seems that a simple pull request update pick up an updated LLVM submodule without any issue. I am not sure why this did not work in this PR. I would suggest that we keep observing our CI and see if this issue arises again. Please copy me if you see CI caching break again in the future, such that I can look more into why certain CI builds behave differently.
